src/fomo.jl

```diff
@@ -0,0 +1,427 @@
+  end
+end
+
+function SolverCore.reset!(solver::FomoSolver{T}) where {T}
+  fill!(solver.m,0)
+  solver
+end
+
+SolverCore.reset!(solver::FomoSolver, ::AbstractNLPModel) = reset!(solver)
+
+function SolverCore.solve!(
+  solver::FomoSolver{T, V},
+  nlp::AbstractNLPModel{T, V},
+  stats::GenericExecutionStats{T, V};
+  callback = (args...) -> nothing,
+  x::V = nlp.meta.x0,
+  atol::T = √eps(T),
+  rtol::T = √eps(T),
+  η1::T = T(eps(T)^(1 / 4)),
+  η2::T = T(0.95),
+  γ1::T = T(1/2),
+  γ2::T = T(2),
+  γ3::T = T(1/2),
+  αmax::T = 1/eps(T),
+  max_time::Float64 = 30.0,
+  max_eval::Int = -1,
+  max_iter::Int = typemax(Int),
+  β::T = T(0.9),
+  θ1::T = T(0.1),
+  θ2::T = T(eps(T)^(1/3)),
+  verbose::Int = 0,
+  backend = r2(),
+  σmin = nothing # keep consistency with R2 interface. kwargs immutable, can't delete it in `R2`
+) where {T, V}
+  r2mode = (backend == R2og())
+  mthname = r2mode ? "R2" : "fomo"
+  unconstrained(nlp) || error("$mthname should only be called on unconstrained problems.")
+  
+  reset!(stats)
+  start_time = time()
+  set_time!(stats, 0.0)
+
+  x = solver.x .= x
+  ∇fk = solver.g
+  c = solver.c
+  m = solver.m
+  d = solver.d
+  set_iter!(stats, 0)
+  set_objective!(stats, obj(nlp, x))
+
+  grad!(nlp, x, ∇fk)
+  norm_∇fk = norm(∇fk)
+  set_dual_residual!(stats, norm_∇fk)
+
+  αk = init_alpha(norm_∇fk,backend)
+  
+  # Stopping criterion: 
+  ϵ = atol + rtol * norm_∇fk
+  optimal = norm_∇fk ≤ ϵ
+  if optimal
+    @info("Optimal point found at initial point")
+    if r2mode
+      @info @sprintf "%5s  %9s  %7s  %7s " "iter" "f" "‖∇f‖" "σ"
+      @info @sprintf "%5d  %9.2e  %7.1e  %7.1e" stats.iter stats.objective norm_∇fk 1/αk
+    else
+      @info @sprintf "%5s  %9s  %7s  %7s " "iter" "f" "‖∇f‖" "α"
+      @info @sprintf "%5d  %9.2e  %7.1e  %7.1e" stats.iter stats.objective norm_∇fk αk
+    end
+    
+  end
+  if verbose > 0 && mod(stats.iter, verbose) == 0
+    if r2mode
+      @info @sprintf "%5s  %9s  %7s  %7s" "iter" "f" "‖∇f‖" "σ"
+    infoline = @sprintf "%5d  %9.2e  %7.1e  %7.1e" stats.iter stats.objective norm_∇fk 1/αk
+    else
+      @info @sprintf "%5s  %9s  %7s  %7s  %7s" "iter" "f" "‖∇f‖" "α" "staβ"
+    infoline = @sprintf "%5d  %9.2e  %7.1e  %7.1e  %7.1e" stats.iter stats.objective norm_∇fk αk 0
+    end
+    
+  end
+
+  set_status!(
+    stats,
+    get_status(
+      nlp,
+      elapsed_time = stats.elapsed_time,
+      optimal = optimal,
+      max_eval = max_eval,
+      iter = stats.iter,
+      max_iter = max_iter,
+      max_time = max_time,
+    ),
+  )
+
+  callback(nlp, solver, stats)
+
+  done = stats.status != :unknown
+
+  d .= ∇fk
+  norm_d = norm_∇fk
+  satβ = T(0)
+  ρk = T(0)
+  avgsatβ = T(0)
+  siter = 0
+  oneT = T(1)
+  while !done
+    λk = step_mult(αk,norm_d,backend)
+    c .= x .- λk .* d
+    uf = x == c # step addition underfow on every dimensions, should happen before αk == 0
+    if r2mode
+      ΔTk = norm_∇fk^2 * λk
+    else  
+      ΔTk = dot(∇fk , d) * λk
```

Comment:
Isn’t it true below that `d == ∇fk` if `r2mode == true`? Perhaps you could have a variable `slope = dot(∇fk, d)` that is used in both cases?

Reply:
This is doable, but it implies computing unnecessarily the dot product in R2 case since `dot(∇fk, d) == norm_∇fk^2` and
`norm_∇fk` is already computed.

Reply:
Yes, but you also compute `norm(d)`. It seems something could be simplified here.

Reply:
The model decrease is now computed in a uniform way:
$$ \Delta Tk = \lambda_k \nabla f(x_k)^T d = \lambda_k \nabla f(x_k)^T [(1-\beta_{max})\nabla f(x_k) + \beta_{max} b_{k-1} = \|\nabla f(x_k)\|^2 + \beta_{max} b_{k-1}^T\nabla f(x_k)$$
There is now a variable `mdot∇f` for $ b_{k-1}^T\nabla f(x_k)$, set to `0` in R2 mode and recomputed when $b_{k-1}$ is updated if momentum is used.